File: packages/flutter_shared/lib/tools/jwt_tool.dart
import 'dart:convert';

import 'jwt_result.dart';

/// JWT 解码工具
class JwtTool {
  JwtTool._();

  /// 解析 JWT token
  static ({JwtResult? result, String? error}) decode(String token) {
    token = token.trim();
    if (token.isEmpty) {
      return (result: null, error: '请输入 JWT Token');
    }

    final parts = token.split('.');
    if (parts.length != 3) {
      return (result: null, error: '非法 JWT 格式：Token 必须由三部分组成（Header.Payload.Signature）');
    }

    try {
      final header = _decodeBase64Url(parts[0]);
      final payload = _decodeBase64Url(parts[1]);
      final signature = parts[2];

      return (
        result: JwtResult(
          header: header,
          payload: payload,
          signature: signature,
        ),
        error: null,
      );
    } catch (e) {
      return (result: null, error: '解码失败: $e');
    }
  }

  static Map<String, dynamic> _decodeBase64Url(String str) {
    // Base64 URL → Base64
    var b64 = str.replaceAll('-', '+').replaceAll('_', '/');
    // 补 = padding
    while (b64.length % 4 != 0) {
      b64 += '=';
    }
    final decoded = base64.decode(b64);
    return jsonDecode(utf8.decode(decoded)) as Map<String, dynamic>;
  }
}

File: miniapp/src/utils/tools/jwtTool.ts
/**
 * JWT 解码工具
 * 移植自 flutter_shared/tools/jwt_tool.dart
 */

export interface JwtResult {
  header: Record<string, any>;
  payload: Record<string, any>;
  signature: string;
}

export interface JwtDecodeResult {
  result?: JwtResult;
  error?: string;
}

function base64UrlDecode(str: string): string {
  let b64 = str.replace(/-/g, '+').replace(/_/g, '/');
  while (b64.length % 4 !== 0) b64 += '=';
  return atob(b64);
}

export const JwtTool = {
  /** 解析 JWT token */
  decode(token: string): JwtDecodeResult {
    token = token.trim();
    if (!token) return { error: '请输入 JWT Token' };

    const parts = token.split('.');
    if (parts.length !== 3) {
      return { error: '非法 JWT 格式：Token 必须由三部分组成（Header.Payload.Signature）' };
    }

    try {
      const headerStr = base64UrlDecode(parts[0]);
      const payloadStr = base64UrlDecode(parts[1]);
      const header = JSON.parse(headerStr);
      const payload = JSON.parse(payloadStr);

      return {
        result: {
          header,
          payload,
          signature: parts[2],
        },
      };
    } catch (e: any) {
      return { error: `解码失败: ${e.message}` };
    }
  },
};
